### PR TITLE
Added first part of unit tests for pkg/detector

### PR DESCRIPTION
**Pull request by anujagrawal699**
**Description:**
This PR significantly enhances the unit tests for the detector package, focusing on the ResourceDetector component. The improvements aim to increase code reliability, maintainability, and test coverage for various detector functions.

**Modifications:**
1. pkg/detector/detector_test.go

**Key Improvements:**
- Implemented tests for GVRDisabled function
- Added tests for NeedLeaderElection and EventFilter function
- Implemented comprehensive tests for OnAdd, OnUpdate, and OnDelete functions
- Added tests for LookForMatchedPolicy and LookForMatchedClusterPolicy functions
- Added tests for ApplyPolicy and ApplyClusterPolicy functions
- Created mock implementations for various interfaces to facilitate testing

Test Coverage:
pkg/detector/detector.go: 1.60% to 32.60%

**What type of PR is this?**
/kind feature

**Which issue(s) this PR fixes**:
Fixes a part of #5470 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Thanks
/assign

File: pkg/detector/detector_test.go [hunk 1]
```diff
@@ -158,3 +180,1092 @@ func BenchmarkEventFilterExtensionApiserverAuthentication(b *testing.B) {
...
 }
+
+func TestGVRDisabled(t *testing.T) {
+	tests := []struct {
+		name     string
+		gvr      schema.GroupVersionResource
+		config   *util.SkippedResourceConfig
+		expected bool
+	}{
+		{
```
Comment (zhzhuang-zju) [inline pkg/detector/detector_test.go:191]: Could you add a test case where `SkippedResourceConfig` is nil? like
```
		{
			name:     "SkippedResourceConfig is nil",
			gvr:      schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments"},
			config:   nil,
			expected: false,
		},
```

**Comment by zhzhuang-zju** [comment]
@anujagrawal699 Good job! others LGTM

**Comment by anujagrawal699** [comment]
> @anujagrawal699 Good job! others LGTM

cc @XiShanYongYe-Chang

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/approve
